<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analysis of single-cell RNA-seq data using a topic model, Part 2: practical implementation • fastTopics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of single-cell RNA-seq data using a topic model, Part 2: practical implementation">
<meta property="og:description" content="fastTopics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fastTopics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6-148</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenslab/fastTopics" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analysis of single-cell RNA-seq data using a
topic model, Part 2: practical implementation</h1>
                        <h4 data-toc-skip class="author">Peter
Carbonetto</h4>
            
            <h4 data-toc-skip class="date">2023-03-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stephenslab/fastTopics/blob/HEAD/vignettes/single_cell_rnaseq_practical.Rmd" class="external-link"><code>vignettes/single_cell_rnaseq_practical.Rmd</code></a></small>
      <div class="hidden name"><code>single_cell_rnaseq_practical.Rmd</code></div>

    </div>

    
    
<p>This vignette continues from <a href="https://stephenslab.github.io/fastTopics/articles/single_cell_rnaseq_basic.html" class="external-link">Part
1</a>, where we introduced the basic steps of a topic modeling analysis
for single-cell data. Here we give more detailed guidance and discuss
complications that may arise.</p>
<p>We begin our analysis by loading the packages. Then we set the seed
so that the results can be reproduced.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://Matrix.R-forge.R-project.org/" class="external-link">Matrix</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stephenslab/fastTopics" class="external-link">fastTopics</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>Now we load the data set and the <span class="math inline">\(K =
6\)</span> pre-fitted topic model.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pbmc_facs</span><span class="op">)</span>
<span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">counts</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">fit</span></code></pre></div>
<div class="section level2">
<h2 id="assessing-convergence-of-model-fitting">Assessing convergence of model fitting<a class="anchor" aria-label="anchor" href="#assessing-convergence-of-model-fitting"></a>
</h2>
<p>The topic model was fitted using the <code>fit_topic_model</code>
function. This function, as we mentioned, hides most of the complexities
of model fitting. Nevertheless, it is a good idea to check that the
model fitting has converged to a maximum-likelihood solution. This is
easily done with ‘fastTopics’:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_progress.html">plot_progress</a></span><span class="op">(</span><span class="va">fit</span>,x <span class="op">=</span> <span class="st">"iter"</span>,add.point.every <span class="op">=</span> <span class="fl">10</span>,colors <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span>font_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><img src="single_cell_rnaseq_practical_files/figure-html/plot-loglik-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Internally, <code>fit_topic_model</code> first fits a non-negative
matrix factorization to the count data, then converts the fitted NMF to
a topic model, so the plot shows the improvement in the model fit over
time. Judging by this plot, the parameter estimates get close to a
maxium-likelihood solution after about 150 updates.</p>
<p>For larger or more complex data sets, more than 200 updates may be
needed to obtain a good fit. For guidance, see
<code><a href="../reference/fit_topic_model.html">help(fit_topic_model)</a></code> and
<code><a href="../reference/fit_poisson_nmf.html">help(fit_poisson_nmf)</a></code>.</p>
</div>
<div class="section level2">
<h2 id="evaluating-fit-single-cell-likelihoods">Evaluating fit: single-cell likelihoods<a class="anchor" aria-label="anchor" href="#evaluating-fit-single-cell-likelihoods"></a>
</h2>
<p>The topic model can be used to calculate <em>a likelihood for each
cell:</em></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">loglik</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/likelihood.html">loglik_multinom_topic_model</a></span><span class="op">(</span><span class="va">counts</span>,<span class="va">fit</span><span class="op">)</span></code></pre></div>
<p>This can be used to assess how well the topic model “fits” each
cell.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">loglik</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pdat</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">loglik</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">64</span>,color <span class="op">=</span> <span class="st">"white"</span>,fill <span class="op">=</span> <span class="st">"black"</span>,size <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"number of cells"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span>font_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><img src="single_cell_rnaseq_practical_files/figure-html/loglik-2-1.png" width="540" style="display: block; margin: auto;"></p>
<p>Most of the poorly fit cells are in the CD34+ subpopulation:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">subpop_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dodgerblue"</span>,<span class="st">"forestgreen"</span>,<span class="st">"darkmagenta"</span>,<span class="st">"skyblue"</span>,<span class="st">"gold"</span><span class="op">)</span>
<span class="va">pdat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>loglik <span class="op">=</span> <span class="va">loglik</span>,subpop <span class="op">=</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">subpop</span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">pdat</span>,<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">loglik</span>,fill <span class="op">=</span> <span class="va">subpop</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>bins <span class="op">=</span> <span class="fl">64</span>,color <span class="op">=</span> <span class="st">"white"</span>,size <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">subpop_colors</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"number of cells"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://wilkelab.org/cowplot/reference/theme_cowplot.html" class="external-link">theme_cowplot</a></span><span class="op">(</span>font_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><img src="single_cell_rnaseq_practical_files/figure-html/loglik-3-1.png" width="660" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="visualizing-the-topics-without-cell-labels">Visualizing the topics without cell labels<a class="anchor" aria-label="anchor" href="#visualizing-the-topics-without-cell-labels"></a>
</h2>
<p>In Part 1, we made use of additional information about the cells to
help us interpret the topics. In some cases, however, we may not have
the benefit of external information to guide our explorations. Without
the cell labels, the Structure plot can still be effective:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">topic_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"skyblue"</span>,<span class="st">"forestgreen"</span>,<span class="st">"darkmagenta"</span>,<span class="st">"dodgerblue"</span>,
                  <span class="st">"gold"</span>,<span class="st">"darkorange"</span><span class="op">)</span>
<span class="fu"><a href="../reference/structure_plot.html">structure_plot</a></span><span class="op">(</span><span class="va">fit</span>,colors <span class="op">=</span> <span class="va">topic_colors</span><span class="op">)</span></code></pre></div>
<p><img src="single_cell_rnaseq_practical_files/figure-html/structure-plot-without-labels-1.png" width="780" style="display: block; margin: auto;"></p>
<p>From this Structure plot, we see clustering into B cells (dark blue),
CD14+ monocytes (green) and CD34+ cells (purple). The clustering of the
NK (light blue) and T cells (mixtures of yellow and orange) is less
distinct, and this is reflected by more sharing of topics (mostly topic
6) among NK and T cells.</p>
<p>Of course, without the cell labels, we cannot know that the topics
correspond to these cell types without further downstream analyses—for
this, we perform a GoM DE analysis as we explained in Part 1.</p>
</div>
<div class="section level2">
<h2 id="identifying-clusters-from-the-mixture-proportions">Identifying clusters from the mixture proportions<a class="anchor" aria-label="anchor" href="#identifying-clusters-from-the-mixture-proportions"></a>
</h2>
<p>In more complex data sets, some structure may not show up well
without additional fine-tuning of the plot.</p>
<p>One simple strategy we have found to work well is to first identify
clusters in the topic proportions matrix, <span class="math inline">\(L\)</span>. Here, we illustrate this idea using
<em>k</em>-means. Specifically, we apply <span class="math inline">\(k\)</span>-means to the principal components (PCs)
of the mixture proportions.</p>
<p>We previously experimented with different numbers of clusters and we
found that 6 clusters worked best. Note that <em>k</em>-means can be
sensitive to (random) initialization, so we set the seed before running
<em>k</em>-means to ensure that the output is reproducible.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">L</span><span class="op">)</span><span class="op">$</span><span class="va">x</span>
<span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/kmeans.html" class="external-link">kmeans</a></span><span class="op">(</span><span class="va">pca</span>,centers <span class="op">=</span> <span class="fl">6</span>,iter.max <span class="op">=</span> <span class="fl">100</span><span class="op">)</span><span class="op">$</span><span class="va">cluster</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span><span class="op">)</span>
<span class="co">#    1    2    3    4    5    6 </span>
<span class="co">#  400  207 1047  797  616  707</span></code></pre></div>
<p>Now we create another Structure plot in which the cells are grouped
by their membership to the 6 clusters:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/structure_plot.html">structure_plot</a></span><span class="op">(</span><span class="va">fit</span>,topics <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,colors <span class="op">=</span> <span class="va">topic_colors</span>,gap <span class="op">=</span> <span class="fl">25</span>,
               grouping <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span></code></pre></div>
<p><img src="single_cell_rnaseq_practical_files/figure-html/structure-plot-by-cluster-1-1.png" width="900" style="display: block; margin: auto;"></p>
<p>Judging by the Structure plot, the <em>k</em>-means clusters look
sensible, the one exception being that the T cells (cells with high
proportions of topics 5 and 6) are arbitrarily split into clusters 1 and
3. Therefore, we merge these two clusters:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clusters</span><span class="op">[</span><span class="va">clusters</span> <span class="op">==</span> <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span></code></pre></div>
<p>We have found that visual inspection of the clusters followed by
manual refinement is often be needed to get the “right” clustering.</p>
<p>For clarity, we label the clusters by these cell types in the plot,
noting that labeling the clusters is usually not possible until
downstream analyses have been performed.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T"</span>,<span class="st">"CD14+"</span>,<span class="st">"B"</span>,<span class="st">"CD34+"</span>,<span class="st">"NK"</span><span class="op">)</span>
<span class="fu"><a href="../reference/structure_plot.html">structure_plot</a></span><span class="op">(</span><span class="va">fit</span>,topics <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,colors <span class="op">=</span> <span class="va">topic_colors</span>,gap <span class="op">=</span> <span class="fl">25</span>,
               grouping <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span></code></pre></div>
<p><img src="single_cell_rnaseq_practical_files/figure-html/structure-plot-by-cluster-2-1.png" width="900" style="display: block; margin: auto;"></p>
<p>For clustering, it is also sometimes helpful to visualize the
clusters in PCA plots which show the projection of the cells onto PCs of
the mixture proportions:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gold"</span>,<span class="st">"forestgreen"</span>,<span class="st">"dodgerblue"</span>,<span class="st">"darkmagenta"</span>,<span class="st">"skyblue"</span><span class="op">)</span>
<span class="fu"><a href="../reference/embedding_plots.html">pca_plot</a></span><span class="op">(</span><span class="va">fit</span>,fill <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cluster_colors</span><span class="op">)</span></code></pre></div>
<p><img src="single_cell_rnaseq_practical_files/figure-html/pca-plot-1-1.png" width="420" style="display: block; margin: auto;"></p>
<p>(The CD14+ and CD34+ cells separate better in PCs 3 and 4.)</p>
<p>When there are many overlapping points, as is the case here, plotting
instead the density of the points can sometimes show the clustering more
clearly:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/embedding_plots.html">pca_hexbin_plot</a></span><span class="op">(</span><span class="va">fit</span>,bins <span class="op">=</span> <span class="fl">24</span><span class="op">)</span></code></pre></div>
<p><img src="single_cell_rnaseq_practical_files/figure-html/pca-plot-2-1.png" width="450" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="more-on-differential-expression-analysis">More on differential expression analysis<a class="anchor" aria-label="anchor" href="#more-on-differential-expression-analysis"></a>
</h2>
<p>In Part 1, we performed a DE analysis using the topic model.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">de</span> <span class="op">&lt;-</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">de</span></code></pre></div>
<p>When the volcano plot shows many overlapping differentially expressed
genes, it can be helpful to explore the results interactively. The
function <code>volcano_plotly</code> creates an <em>interactive volcano
plot</em> that can be viewed in a Web browser. For example, here is an
interactive volcano plot for the T cells topic:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">pbmc_facs</span><span class="op">$</span><span class="va">genes</span>
<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/volcano_plot.html">volcano_plotly</a></span><span class="op">(</span><span class="va">de</span>,k <span class="op">=</span> <span class="fl">6</span>,file <span class="op">=</span> <span class="st">"volcano_plot_t_cells.html"</span>,
                    labels <span class="op">=</span> <span class="va">genes</span><span class="op">$</span><span class="va">symbol</span>,ymax <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></code></pre></div>
<p>This call creates a new file <code>volcano_plot_t_cells.html</code>.
The interactive volcano plot can also be viewed <a href="https://stephenslab.github.io/fastTopics/articles/volcano_plot_t_cells.html" class="external-link">here</a>,
or by calling <code>print(p)</code> in R.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<p>This is the version of R and the packages that were used to generate
these results.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># R version 3.6.2 (2019-12-12)</span>
<span class="co"># Platform: x86_64-apple-darwin15.6.0 (64-bit)</span>
<span class="co"># Running under: macOS Catalina 10.15.7</span>
<span class="co"># </span>
<span class="co"># Matrix products: default</span>
<span class="co"># BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib</span>
<span class="co"># LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib</span>
<span class="co"># </span>
<span class="co"># locale:</span>
<span class="co"># [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co"># </span>
<span class="co"># attached base packages:</span>
<span class="co"># [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co"># </span>
<span class="co"># other attached packages:</span>
<span class="co"># [1] cowplot_1.1.1      ggplot2_3.3.6      fastTopics_0.6-147 Matrix_1.2-18     </span>
<span class="co"># </span>
<span class="co"># loaded via a namespace (and not attached):</span>
<span class="co">#  [1] mcmc_0.9-6         fs_1.5.2           progress_1.2.2     httr_1.4.2        </span>
<span class="co">#  [5] rprojroot_1.3-2    tools_3.6.2        backports_1.1.5    bslib_0.3.1       </span>
<span class="co">#  [9] utf8_1.1.4         R6_2.4.1           irlba_2.3.3        uwot_0.1.10       </span>
<span class="co"># [13] DBI_1.1.0          lazyeval_0.2.2     colorspace_1.4-1   withr_2.5.0       </span>
<span class="co"># [17] tidyselect_1.1.1   prettyunits_1.1.1  compiler_3.6.2     cli_3.5.0         </span>
<span class="co"># [21] quantreg_5.54      SparseM_1.78       desc_1.2.0         plotly_4.10.1     </span>
<span class="co"># [25] labeling_0.3       sass_0.4.0         scales_1.1.0       SQUAREM_2017.10-1 </span>
<span class="co"># [29] hexbin_1.28.0      quadprog_1.5-8     pbapply_1.5-1      pkgdown_2.0.2     </span>
<span class="co"># [33] mixsqp_0.3-46      systemfonts_1.0.2  stringr_1.4.0      digest_0.6.23     </span>
<span class="co"># [37] rmarkdown_2.11     MCMCpack_1.4-5     pkgconfig_2.0.3    htmltools_0.5.4   </span>
<span class="co"># [41] fastmap_1.1.0      invgamma_1.1       highr_0.8          htmlwidgets_1.6.1 </span>
<span class="co"># [45] rlang_1.0.6        shiny_1.7.4        jquerylib_0.1.4    generics_0.0.2    </span>
<span class="co"># [49] farver_2.0.1       jsonlite_1.7.2     crosstalk_1.0.0    dplyr_1.0.7       </span>
<span class="co"># [53] magrittr_2.0.1     Rcpp_1.0.8         munsell_0.5.0      fansi_0.4.0       </span>
<span class="co"># [57] lifecycle_1.0.3    stringi_1.4.3      yaml_2.2.0         MASS_7.3-51.4     </span>
<span class="co"># [61] Rtsne_0.15         grid_3.6.2         promises_1.1.0     parallel_3.6.2    </span>
<span class="co"># [65] ggrepel_0.9.1      crayon_1.4.1       lattice_0.20-38    hms_1.1.0         </span>
<span class="co"># [69] knitr_1.37         pillar_1.6.2       glue_1.4.2         evaluate_0.14     </span>
<span class="co"># [73] data.table_1.12.8  RcppParallel_5.1.5 httpuv_1.5.2       vctrs_0.3.8       </span>
<span class="co"># [77] MatrixModels_0.4-1 gtable_0.3.0       purrr_0.3.4        tidyr_1.1.3       </span>
<span class="co"># [81] assertthat_0.2.1   ashr_2.2-54        xfun_0.29          mime_0.8          </span>
<span class="co"># [85] xtable_1.8-4       later_1.0.0        coda_0.19-3        ragg_0.3.1        </span>
<span class="co"># [89] viridisLite_0.3.0  truncnorm_1.0-8    tibble_3.1.3       memoise_1.1.0     </span>
<span class="co"># [93] ellipsis_0.3.2</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Peter Carbonetto, Kevin Luo, Kushal Dey, Matthew
Stephens.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
