<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Topic modeling vs. clustering of gene expression data: an illustration • fastTopics</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Topic modeling vs. clustering of gene expression data: an illustration">
<meta property="og:description" content="fastTopics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fastTopics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5-52</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenslab/fastTopics">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Topic modeling vs. clustering of gene expression data: an illustration</h1>
                        <h4 class="author">Peter Carbonetto</h4>
            
            <h4 class="date">2021-06-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stephenslab/fastTopics/blob/master/vignettes/topics_vs_clusters.Rmd"><code>vignettes/topics_vs_clusters.Rmd</code></a></small>
      <div class="hidden name"><code>topics_vs_clusters.Rmd</code></div>

    </div>

    
    
<p>Clustering methods are widely used to discover interesting substructure in bulk or single-cell RNA sequencing (“RNA-seq”) data. An alternative is to use a topic model, which explains each sample (an expression profile) as a mixture of “gene programs” that are estimated from the data. In a topic modeling analysis, the gene expression programs are the topics.</p>
<p>In this short vignette, we show that a topic model picks up a very different kind of substructure that cannot be identified by (hard) clustering methods.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">cowplot</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Ternary</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Rtsne</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">fastTopics</span>)</pre></body></html></div>
<p>Simulate a “toy” gene expression data set from the multinomial topic model. Each sample—here, a vector of 400 counts—is generated as a mixture of three gene programs (the topics in the topic model). In particular, samples are mostly generated by either a single gene program or an approximately equal mixture of the three gene programs.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1</span>)
<span class="no">X</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/simulate_toy_gene_data.html">simulate_toy_gene_data</a></span>(<span class="kw">n</span> <span class="kw">=</span> <span class="fl">400</span>,<span class="kw">m</span> <span class="kw">=</span> <span class="fl">40</span>,<span class="kw">k</span> <span class="kw">=</span> <span class="fl">3</span>,<span class="kw">s</span> <span class="kw">=</span> <span class="fl">1000</span>)$<span class="no">X</span></pre></body></html></div>
<div id="topic-modeling-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#topic-modeling-analysis" class="anchor"></a>Topic modeling analysis</h2>
<p>To illustrate a topic modeling analysis, we fit a multinomial topic model to these data,</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fit_poisson_nmf.html">fit_poisson_nmf</a></span>(<span class="no">X</span>,<span class="kw">k</span> <span class="kw">=</span> <span class="fl">3</span>,<span class="kw">numiter</span> <span class="kw">=</span> <span class="fl">100</span>,<span class="kw">verbose</span> <span class="kw">=</span> <span class="st">"none"</span>,
                       <span class="kw">control</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="kw">extrapolate</span> <span class="kw">=</span> <span class="fl">TRUE</span>))
<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/poisson2multinom.html">poisson2multinom</a></span>(<span class="no">fit</span>)</pre></body></html></div>
<p>then we plot the estimated topic proportions.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="kw">mar</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>))
<span class="no">pdat</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">fit</span>$<span class="no">L</span>)
<span class="fu"><a href="https://ms609.github.io/Ternary/reference/TernaryPlot.html">TernaryPlot</a></span>(<span class="kw">alab</span> <span class="kw">=</span> <span class="st">"k1"</span>,<span class="kw">blab</span> <span class="kw">=</span> <span class="st">"k2"</span>,<span class="kw">clab</span> <span class="kw">=</span> <span class="st">"k3"</span>,<span class="kw">grid.col</span> <span class="kw">=</span> <span class="st">"skyblue"</span>,
            <span class="kw">grid.minor.lines</span> <span class="kw">=</span> <span class="fl">0</span>)
<span class="fu"><a href="https://ms609.github.io/Ternary/reference/AddToTernary.html">TernaryPoints</a></span>(<span class="no">pdat</span>,<span class="kw">pch</span> <span class="kw">=</span> <span class="fl">21</span>,<span class="kw">col</span> <span class="kw">=</span> <span class="st">"white"</span>,<span class="kw">bg</span> <span class="kw">=</span> <span class="st">"black"</span>,<span class="kw">cex</span> <span class="kw">=</span> <span class="fl">0.8</span>)</pre></body></html></div>
<p><img src="topics_vs_clusters_files/figure-html/plot-topic-proportions-1.png" width="360" style="display: block; margin: auto;"></p>
<p>We observe in this plot that the estimated topic model recaptures the simulation: according to the model, samples are primarily explained by a single gene program, or by an approximately equal mixture of three gene programs.</p>
</div>
<div id="clustering-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering-analysis" class="anchor"></a>Clustering analysis</h2>
<p>To illustrate a clustering analysis, we use the <em>t</em>-SNE algorithm to project the samples onto a 2-d embedding.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">tsne1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html">Rtsne</a></span>(<span class="no">X</span>,<span class="fl">2</span>,<span class="kw">pca</span> <span class="kw">=</span> <span class="fl">FALSE</span>,<span class="kw">normalize</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tsne1</span>$<span class="no">Y</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"d1"</span>,<span class="st">"d2"</span>)
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">tsne1</span>$<span class="no">Y</span>),<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">d1</span>,<span class="kw">y</span> <span class="kw">=</span> <span class="no">d2</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw">shape</span> <span class="kw">=</span> <span class="fl">21</span>,<span class="kw">color</span> <span class="kw">=</span> <span class="st">"white"</span>,<span class="kw">fill</span> <span class="kw">=</span> <span class="st">"black"</span>,<span class="kw">size</span> <span class="kw">=</span> <span class="fl">2</span>) +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/theme_cowplot.html">theme_cowplot</a></span>(<span class="kw">font_size</span> <span class="kw">=</span> <span class="fl">12</span>)</pre></body></html></div>
<p><img src="topics_vs_clusters_files/figure-html/tsne-from-counts-1-1.png" width="360" style="display: block; margin: auto;"></p>
<p>This 2-d projection immediately suggests four clusters. (In practice, one could automate the clustering of these data using a community detection method such as the Louvain algorithm, or by applying <span class="math inline">\(k\)</span>-means to the <span class="math inline">\(t\)</span>-SNE projection.)</p>
<p>Although the clustering is correct—there are indeed four clusters—it fails to reveal any insight into the relationship between these clusters. In particular, the distribution of genes in the middle cluster is an equal mixture of the gene distributions in the other clusters. This is a limitation of any clustering method that does not allow for partial membership to clusters.</p>
<p>To further understand this point, we label each sample by the topic with the highest estimated weight:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">kmax</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="no">fit</span>$<span class="no">L</span>,<span class="fl">1</span>,<span class="no">which.max</span>))
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">tsne1</span>$<span class="no">Y</span>,<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">kmax</span>)),<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">d1</span>,<span class="kw">y</span> <span class="kw">=</span> <span class="no">d2</span>,<span class="kw">fill</span> <span class="kw">=</span> <span class="no">kmax</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw">shape</span> <span class="kw">=</span> <span class="fl">21</span>,<span class="kw">color</span> <span class="kw">=</span> <span class="st">"white"</span>,<span class="kw">size</span> <span class="kw">=</span> <span class="fl">2</span>,<span class="kw">show.legend</span> <span class="kw">=</span> <span class="fl">FALSE</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dodgerblue"</span>,<span class="st">"darkorange"</span>,<span class="st">"darkblue"</span>)) +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/theme_cowplot.html">theme_cowplot</a></span>(<span class="kw">font_size</span> <span class="kw">=</span> <span class="fl">12</span>)</pre></body></html></div>
<p><img src="topics_vs_clusters_files/figure-html/tsne-from-counts-2-1.png" width="360" style="display: block; margin: auto;"></p>
<p>In general, a clustering will align closely with the topic modeling whenever the samples can be largely explained by a single topic or gene program. Outside this setting, topic modeling and clustering can reveal very different sorts of substructure, so a topic modeling analysis should not be expected to recapture the results of a clustering analysis, and vice versa.</p>
</div>
<div id="inferring-clusters-from-the-topic-model" class="section level2">
<h2 class="hasAnchor">
<a href="#inferring-clusters-from-the-topic-model" class="anchor"></a>Inferring clusters from the topic model</h2>
<p>Despite this limitation of the clustering methods, it can nonetheless be useful to identify the clusters exhibiting markedly different gene expression patterns. This can also be done within a topic modeling analysis: the topic model contains enough information to identify these clusters, at least in the artificial setting where the data are simulated from a topic model.</p>
<p>For example, a 2-d <em>t</em>-SNE projection of the topic proportions reveals the same four clusters:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">tsne2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rtsne/man/Rtsne.html">Rtsne</a></span>(<span class="no">fit</span>$<span class="no">L</span>,<span class="fl">2</span>,<span class="kw">pca</span> <span class="kw">=</span> <span class="fl">FALSE</span>,<span class="kw">normalize</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tsne2</span>$<span class="no">Y</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"d1"</span>,<span class="st">"d2"</span>)
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">tsne2</span>$<span class="no">Y</span>,<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">kmax</span>)),<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">d1</span>,<span class="kw">y</span> <span class="kw">=</span> <span class="no">d2</span>,<span class="kw">fill</span> <span class="kw">=</span> <span class="no">kmax</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw">shape</span> <span class="kw">=</span> <span class="fl">21</span>,<span class="kw">color</span> <span class="kw">=</span> <span class="st">"white"</span>,<span class="kw">size</span> <span class="kw">=</span> <span class="fl">2</span>,<span class="kw">show.legend</span> <span class="kw">=</span> <span class="fl">FALSE</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dodgerblue"</span>,<span class="st">"darkorange"</span>,<span class="st">"darkblue"</span>)) +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/theme_cowplot.html">theme_cowplot</a></span>(<span class="kw">font_size</span> <span class="kw">=</span> <span class="fl">12</span>)</pre></body></html></div>
<p><img src="topics_vs_clusters_files/figure-html/tsne-from-loadings-1.png" width="360" style="display: block; margin: auto;"></p>
<p>Or we can achieve a similar result using PCA:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">pca</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="no">fit</span>$<span class="no">L</span>,<span class="kw">center</span> <span class="kw">=</span> <span class="fl">TRUE</span>,<span class="kw">scale</span> <span class="kw">=</span> <span class="fl">FALSE</span>)
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">pca</span>$<span class="no">x</span>,<span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="no">kmax</span>)),<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">PC1</span>,<span class="kw">y</span> <span class="kw">=</span> <span class="no">PC2</span>,<span class="kw">fill</span> <span class="kw">=</span> <span class="no">kmax</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw">shape</span> <span class="kw">=</span> <span class="fl">21</span>,<span class="kw">color</span> <span class="kw">=</span> <span class="st">"white"</span>,<span class="kw">size</span> <span class="kw">=</span> <span class="fl">2</span>,<span class="kw">show.legend</span> <span class="kw">=</span> <span class="fl">FALSE</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span>(<span class="kw">values</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"dodgerblue"</span>,<span class="st">"darkorange"</span>,<span class="st">"darkblue"</span>)) +
  <span class="fu"><a href="https://rdrr.io/pkg/cowplot/man/theme_cowplot.html">theme_cowplot</a></span>(<span class="kw">font_size</span> <span class="kw">=</span> <span class="fl">12</span>)</pre></body></html></div>
<p><img src="topics_vs_clusters_files/figure-html/pca-from-loadings-1.png" width="360" style="display: block; margin: auto;"></p>
<p>In short, the topic modeling analysis retains the benefits of the clustering analysis, while revealing additional substructure that cannot be identified by standard clustering methods.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Peter Carbonetto, Kevin Luo, Kushal Dey, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
