<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Differential Count Analysis with a Multinomial Topic Model — diff_count_analysis • fastTopics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Differential Count Analysis with a Multinomial Topic Model — diff_count_analysis" />
<meta property="og:description" content="Implements methods for analysis of differential count
analysis using a topic model. These methods are motivated by gene
expression studies, but could have other uses, such as identifying
&amp;#8220;key words&amp;#8221; in topics derived from text documents. To
improve accuracy of the differential expression analysis, an
empirical Bayes method is used to &amp;#8220;stabilize&amp;#8221; the
estimates. A special case of &amp;#8220;hard&amp;#8221; topic assignments is
also implemented---that is, the mixture proportions are all zeros
and ones---which involves greatly simplified (and faster)
calculations. Use diff_count_clusters for this special case." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fastTopics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5-52</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/fastTopics">Source</a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Differential Count Analysis with a Multinomial Topic Model</h1>
    <small class="dont-index">Source: <a href='https://github.com/stephenslab/fastTopics/blob/master/R/diffcount.R'><code>R/diffcount.R</code></a></small>
    <div class="hidden name"><code>diff_count_analysis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Implements methods for analysis of differential count
analysis using a topic model. These methods are motivated by gene
expression studies, but could have other uses, such as identifying
&#8220;key words&#8221; in topics derived from text documents. To
improve accuracy of the differential expression analysis, an
empirical Bayes method is used to &#8220;stabilize&#8221; the
estimates. A special case of &#8220;hard&#8221; topic assignments is
also implemented---that is, the mixture proportions are all zeros
and ones---which involves greatly simplified (and faster)
calculations. Use <code>diff_count_clusters</code> for this special case.</p>
    </div>

    <pre class="usage"><span class='fu'>diff_count_analysis</span>(
  <span class='no'>fit</span>,
  <span class='no'>X</span>,
  <span class='kw'>s</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowSums</a></span>(<span class='no'>X</span>),
  <span class='kw'>pseudocount</span> <span class='kw'>=</span> <span class='fl'>0.01</span>,
  <span class='kw'>fit.method</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"em"</span>, <span class='st'>"optim"</span>, <span class='st'>"glm"</span>),
  <span class='kw'>shrink.method</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"ash"</span>, <span class='st'>"none"</span>),
  <span class='kw'>numiter</span> <span class='kw'>=</span> <span class='fl'>100</span>,
  <span class='kw'>tol</span> <span class='kw'>=</span> <span class='fl'>1e-08</span>,
  <span class='kw'>e</span> <span class='kw'>=</span> <span class='fl'>1e-15</span>,
  <span class='kw'>show.warning</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='no'>...</span>
)

<span class='fu'>diff_count_clusters</span>(<span class='no'>cluster</span>, <span class='no'>X</span>, <span class='no'>...</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>fit</th>
      <td><p>An object of class &#8220;poisson_nmf_fit&#8221; or
&#8220;multinom_topic_model_fit&#8221;. If a Poisson NMF fit is provided
as input, the corresponding multinomial topic model fit is
automatically recovered using <code><a href='poisson2multinom.html'>poisson2multinom</a></code>.</p></td>
    </tr>
    <tr>
      <th>X</th>
      <td><p>The n x m counts matrix. It can be a sparse matrix (class
<code>"dgCMatrix"</code>) or dense matrix (class <code>"matrix"</code>).</p></td>
    </tr>
    <tr>
      <th>s</th>
      <td><p>A numeric vector of length n determining how the rates are
scaled in the Poisson model. See &#8220;Details&#8221; for guidance on
the choice of <code>s</code>.</p></td>
    </tr>
    <tr>
      <th>pseudocount</th>
      <td><p>Observations with this value are added to the
counts matrix to stabilize calculation of the LFC estimates and
other statistics.</p></td>
    </tr>
    <tr>
      <th>fit.method</th>
      <td><p>Method used to estimate LFC and compute test
statistics. The <code>"glm"</code> and <code>"optim"</code> computations are
particularly slow, and it is recommended to use <code>fit.method =
"em"</code> for most data sets. See &#8220;Details&#8221; for more
information.</p></td>
    </tr>
    <tr>
      <th>shrink.method</th>
      <td><p>Method used to stabilize the LFC estimates.
When <code>shrink.method = "ash"</code>, the "adaptive shrinkage" method
implemented in the ashr package is used. When <code>shrink.method =
"none"</code>, no stabilization is performed, and the &#8220;raw&#8221; LFC
estimates are returned.</p></td>
    </tr>
    <tr>
      <th>numiter</th>
      <td><p>Maximum number of iterations performed in
optimization of the Poisson model parameters. When
<code>fit.method = "glm"</code>, this is passed as argument <code>maxit</code>
to the <code>glm</code> function; when <code>fit.method = "optim"</code>, this
is passed as argument <code>maxit</code> to the <code>optim</code> function.</p></td>
    </tr>
    <tr>
      <th>tol</th>
      <td><p>Controls the convergence tolerance for the optimization
of the Poisson model parameters. When <code>fit.method = "glm"</code>,
this is passed as argument <code>epsilon</code> to function <code>glm</code>;
when <code>fit.method = "optim"</code>, the <code>factr</code> optimization
setting is set to <code>tol * .Machine$double.eps</code>. When
<code>fit.method = "em"</code>, the EM algorithm will be stopped early
when the largest change between two successive updates is less than
<code>tol</code>.</p></td>
    </tr>
    <tr>
      <th>e</th>
      <td><p>A small, positive scalar included in some computations to
avoid logarithms of zero and division by zero.</p></td>
    </tr>
    <tr>
      <th>show.warning</th>
      <td><p>Set <code>show.warning = FALSE</code> to suppress a
message about calculations when mixture proportions are all 0 or 1.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>When <code>verbose = TRUE</code>, progress information is
printed to the console.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments passed to
<code>diff_count_analysis</code>.</p></td>
    </tr>
    <tr>
      <th>cluster</th>
      <td><p>A factor, or vector that can be converted to a
factor such as an integer or character vector, giving an assignment
of samples (rows of <code>X</code>) to clusters. This could be, for
example, the "cluster" output from <code><a href='https://rdrr.io/r/stats/kmeans.html'>kmeans</a></code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The return value is a list of m x k matrices, where m is
  the number of columns in the counts matrix, and k is the number of
  topics (for <code>diff_count_clusters</code>, m is the number of
  clusters), and an additional vector:</p>
<dt>colmeans</dt><dd><p>A vector of length m containing the count averages
  (<code><a href='https://rdrr.io/r/base/colSums.html'>colMeans(X)</a></code>).</p></dd>

<dt>F0</dt><dd><p>Estimates of the Poisson model parameters \(f_0\).</p></dd>

<dt>F1</dt><dd><p>Estimates of the Poisson model parameters \(f_1\).</p></dd>

<dt>beta</dt><dd><p>LFC estimates <code>beta = log2(F1/F0)</code>.</p></dd>

<dt>se</dt><dd><p>Standard errors of the Poisson glm parameters \(b =
  f_1 - f_0\).</p></dd>

<dt>Z</dt><dd><p>z-scores for the Poisson glm parameters \(b = f_1 -
  f_0\).</p></dd>

<dt>pval</dt><dd><p>-log10 two-tailed p-values computed from the
  z-scores. In some extreme cases the calculations may produce zero
  or negative standard errors, in which case the standard errors are
  set to <code>NA</code> and the z-scores and -log10 p-values are set to
  zero.</p></dd>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The methods are based on the following univariate
(&#8220;single-count&#8221;) Poisson model: $$x_i ~ Poisson(s_i
\lambda_i),$$ in which the Poisson rates are defined as
$$\lambda_i = (1 - q_i) f_0 + q_i f_1,$$ and where the two
unknowns to be estimated are \(f_0, f_1 &gt; 0.\) This model is
applied separately to each count (column of the counts matrix), and
to each topic k. The \(q_i\)'s are the topic probabilities for a
given topic. An EM algorithm is used to compute maximum-likelihood
estimates (MLEs) of the two unknowns.</p>
<p>The log-fold change (LFC) statistics are defined as the log-ratio
of the two Poisson rate parameters, \(\beta = \log_2(f_1/f_0)\).
This statistic measures the increase (or decrease) in occurance in
one topic compared to all other topics. The use of the base-2
logarithm comes from the convention used in gene expression
studies.</p>
<p>When each \(s_i\) (specified by input argument <code>s</code>) is equal
the total count for sample i (this is the default setting in
<code>diff_count_analysis</code>), the Poisson model will closely
approximate a binomial model of the count data, so that the Poisson
model parameters \(f_0, f_1\) represent binomial probabilities.
In this case, \(\beta\) represents the LFC in <em>relative</em>
occurrence. (This Poisson approximation to the binomial is most
accurate when the total counts <code><a href='https://rdrr.io/r/base/colSums.html'>rowSums(X)</a></code> are large and
\(f_0, f_1\) are small.)</p>
<p>Other choices for <code>s</code> are possible, and implement different
normalization schemes for the counts, and different interpretations
of the LFC statistics. Setting <code>s</code> to all ones implements the
differential count analysis with no normalization, and \(\beta\)
represents the LFC in <em>absolute</em> occurrence. This choice could
be appropriate in settings where the total count is well-controlled
across samples (rows of <code>X</code>).</p>
<p>When <code>fit.method = "glm"</code>, the LFC estimates and test
statistics are implemented by <code><a href='https://rdrr.io/r/stats/glm.html'>glm</a></code> with
<code>formula = x ~ b0 + b - 1</code> and <code>family = poisson(link =
"identity")</code>, where the unknowns are defined as \(b_0 = f_0\) and
\(b = f_1 - f_0\). The outputted z-scores are p-values are for
the <code>b</code> coefficient; that is, they are test statistics for the
test that \(f_0\) and \(f_1\) are not equal.</p>
<p>When <code>fit.method = "optim"</code> or <code>fit.method = "em"</code>,
maximum-likelihood estimates are computed using
<code><a href='https://rdrr.io/r/stats/optim.html'>optim</a></code> or using a fast EM algorithm, and test
statistics (standard errors, z-scores and p-values) are based on a
Laplace approximation to the likelihood at the MLE. These
calculations closely reproduce the test statistic calculations in
<code><a href='https://rdrr.io/r/stats/glm.html'>glm</a></code>.</p>
<p>We recommend setting <code>shrink.method = "ash"</code>, which uses the
&#8220;adaptive shrinkage&#8221; method (Stephens, 2016) to improve
accuracy of the LFC estimates. The improvement in accuracy is
greatest for genes with low expression, or when the sample size is
small (or both). We follow the settings used in <code>lfcShrink</code>
from the DESeq2 package, with <code>type = "ashr"</code>.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Stephens, M. (2016). False discovery rates: a new deal.
<em>Biostatistics</em> <b>18</b>(2), kxw041.
<a href='https://doi.org/10.1093/biostatistics/kxw041'>https://doi.org/10.1093/biostatistics/kxw041</a></p>
<p>Zhu, A., Ibrahim, J. G. and Love, M. I. (2019). Heavy-tailed prior
distributions for sequence count data: removing the noise and
preserving large differences. <em>Bioinformatics</em> <b>35</b>(12),
2084–2092.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Peter Carbonetto, Kevin Luo, Kushal Dey, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


