<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform HOMER Motif Enrichment Analysis using DE Genomic Positions — run_homer • fastTopics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform HOMER Motif Enrichment Analysis using DE Genomic Positions — run_homer"><meta property="og:description" content="Run HOMER motif finding algorithm
  (findMotifsGenome.pl) to identify motifs enriched for
  differentially expressed (DE) genomic positions. See
  http://homer.ucsd.edu for more information."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fastTopics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6-147</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stephenslab/fastTopics" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform HOMER Motif Enrichment Analysis using DE Genomic Positions</h1>
    <small class="dont-index">Source: <a href="https://github.com/stephenslab/fastTopics/blob/HEAD/R/homer.R" class="external-link"><code>R/homer.R</code></a></small>
    <div class="hidden name"><code>run_homer.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run HOMER motif finding algorithm
  (<code>findMotifsGenome.pl</code>) to identify motifs enriched for
  differentially expressed (DE) genomic positions. See
  <a href="http://homer.ucsd.edu" class="external-link">http://homer.ucsd.edu</a> for more information.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">run_homer</span><span class="op">(</span>
  <span class="va">de</span>,
  <span class="va">k</span>,
  <span class="va">positions</span>,
  genome <span class="op">=</span> <span class="st">"hg19"</span>,
  subset <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">postmean</span>, <span class="va">lpval</span>, <span class="va">lfsr</span>, <span class="va">rank</span>, <span class="va">quantile</span><span class="op">)</span> <span class="va">lfsr</span> <span class="op">&lt;</span> <span class="fl">0.05</span>,
  homer.exec <span class="op">=</span> <span class="st">"findMotifsGenome.pl"</span>,
  out.dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,
  homer.options <span class="op">=</span> <span class="st">"-len 8,10,12 -size 200 -mis 2 -S 25 -p 1 -h"</span>,
  verbose <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>de</dt>
<dd><p>An object of class “topic_model_de_analysis”,
usually the result of running <code><a href="de_analysis.html">de_analysis</a></code>.</p></dd>
<dt>k</dt>
<dd><p>Use the DE analysis results for this topic.</p></dd>
<dt>positions</dt>
<dd><p>A table of genomic positions corresponding to rows
of the <code>de_analysis</code> results. Specifically, it should a data
frame with four columns: “chr”, chromosome name or number;
“start”, start position of genomic feature; “end”,
end position of genomic feature; and “name”, the name of the
genomic feature. If not specified, the genomic positions will be
extracted from the row names of <code>de$postmean</code>, in which the
row names are expected to be of the form <code>chr_start_end</code>. The
genomic positions will be written to a BED file (see
<a href="https://genome.ucsc.edu/FAQ/FAQformat.html" class="external-link">https://genome.ucsc.edu/FAQ/FAQformat.html</a> for more
information about BED files).</p></dd>
<dt>genome</dt>
<dd><p>The genome parameter passed to
<code>findMotifsGenome.pl</code>.</p></dd>
<dt>subset</dt>
<dd><p>Describe input argument "subset" here.</p></dd>
<dt>homer.exec</dt>
<dd><p>The name or file path of the HOMER
<code>findMotifsGenome.pl</code> excutable.</p></dd>
<dt>out.dir</dt>
<dd><p>The positions BED file and HOMER results are written
to this directory.</p></dd>
<dt>homer.options</dt>
<dd><p>Character string used to override default
<code>findMotifsGenome.pl</code> options.</p></dd>
<dt>verbose</dt>
<dd><p>When <code>verbose = TRUE</code>, progress information is
printed to the console.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data frame containing the motif enrichment results. It
  is created from the <code>knownResults.txt</code> HOMER output.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Heinz, S., Benner, C., Spann, N., Bertolino, E., Lin, Y. C., Laslo,
P., Cheng, J. X., Murre, C., Singh, H. and Glass, C. K. (2010).
Simple combinations of lineage-determining transcription factors
prime cis-regulatory elements required for macrophage and B cell
identities. <em>Molecular Cell</em> <b>38</b>, 576-589.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Peter Carbonetto, Kevin Luo, Kushal Dey, Matthew
Stephens.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

