<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Simulate Gene Expression Data from Poisson NMF or Multinomial
  Topic Model — simulate_poisson_gene_data • fastTopics</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate Gene Expression Data from Poisson NMF or Multinomial
  Topic Model — simulate_poisson_gene_data"><meta property="og:description" content="Simulate count data from a Poisson NMF model or
  multinomial topic model, in which topics represent &amp;#8220;gene
  expression programs&amp;#8221;, and gene expression programs are
  characterized by different rates of expression. The way in which
  the counts are simulated is modeled after gene expression studies
  in which expression is measured by single-cell RNA sequencing
  (&amp;#8220;RNA-seq&amp;#8221;) techniques: each row of the counts matrix
  corresponds a gene expression profile, each column corresponds to a
  gene, and each matrix element is a &amp;#8220;read count&amp;#8221;, or
  &amp;#8220;UMI count&amp;#8221;, measuring expression level. Factors are
  simulated so as to capture realistic changes in gene expression
  across different cell types. See &amp;#8220;Details&amp;#8221; for the procedure
  used to simulate factors, loadings and counts."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fastTopics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6-147</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/stephenslab/fastTopics" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate Gene Expression Data from Poisson NMF or Multinomial
  Topic Model</h1>
    <small class="dont-index">Source: <a href="https://github.com/stephenslab/fastTopics/blob/HEAD/R/datasim.R" class="external-link"><code>R/datasim.R</code></a></small>
    <div class="hidden name"><code>simulate_gene_data.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Simulate count data from a Poisson NMF model or
  multinomial topic model, in which topics represent “gene
  expression programs”, and gene expression programs are
  characterized by different rates of expression. The way in which
  the counts are simulated is modeled after gene expression studies
  in which expression is measured by single-cell RNA sequencing
  (“RNA-seq”) techniques: each row of the counts matrix
  corresponds a gene expression profile, each column corresponds to a
  gene, and each matrix element is a “read count”, or
  “UMI count”, measuring expression level. Factors are
  simulated so as to capture realistic changes in gene expression
  across different cell types. See “Details” for the procedure
  used to simulate factors, loadings and counts.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">simulate_poisson_gene_data</span><span class="op">(</span><span class="va">n</span>, <span class="va">m</span>, <span class="va">k</span>, <span class="va">s</span>, p <span class="op">=</span> <span class="fl">1</span>, sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="fu">simulate_multinom_gene_data</span><span class="op">(</span><span class="va">n</span>, <span class="va">m</span>, <span class="va">k</span>, sparse <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>n</dt>
<dd><p>Number of rows in the simulated count matrix. Should be at
least 2.</p></dd>
<dt>m</dt>
<dd><p>Number of columns in the simulated count matrix. Should be
at least 2.</p></dd>
<dt>k</dt>
<dd><p>Number of factors, or “topics”, used to generate
the data. Should be 2 or more.</p></dd>
<dt>s</dt>
<dd><p>Vector of “size factors”; each row of the loadings
matrix <code>L</code> is scaled by the entries of <code>s</code> before
generating the counts. This should be a vector of length n
containing only positive values.</p></dd>
<dt>p</dt>
<dd><p>Probability that <code>F[i,j]</code> is equal to the mean rate.
Smaller values of <code>p</code> will result in more factors that are the
same across topics.</p></dd>
<dt>sparse</dt>
<dd><p>If <code>sparse = TRUE</code>, convert the counts matrix to
a sparse matrix in compressed, column-oriented format; see
<code><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><code>simulate_poisson_gene_data</code> returns a list containing
  the counts matrix <code>X</code>, and the size factors <code>s</code> and
  factorization, <code>F</code>, <code>L</code>, used to generate the counts.
<code>simulate_multinom_gene_data</code> returns a list containing the
  counts matrix <code>X</code>, and the mixture proportions <code>L</code> and
  factors (gene probabilities, or relative gene expression levels)
<code>F</code> used to generate the counts.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Here we describe the process for generating the n x k
  loadings matrix <code>L</code> and the m x k factors matrix <code>F</code>.</p>
<p>Each row of the <code>L</code> matrix is generated in the following
  manner: (1) the number of nonzero mixture proportions is \(1
  \le n \le k\), with probability proportional to \(2^{-n}\);
  (2) the indices of the nonzero mixture proportions are sampled
  uniformly at random; and (3) the nonzero mixture proportions are
  sampled from the Dirichlet distribution with \(\alpha = 1\) (so
  that all topics are equally likely).</p>
<p>Each row of the factors matrix are generated according to the
  following procedure: (1) generate \(u = |r| - 5\), where \(r ~
  N(0,2)\); (2) for each topic \(k\), generate the Poisson rates as
  \(exp(max(t,-5))\), where \(t ~ 0.95 * N(u,s/10) + 0.05 *
  N(u,s)\), and \(s = exp(-u/8)\). Factors can be interpreted as
  Poisson rates or multinomial probabilities, so that individual
  counts can be viewed as being generated from a weighted mixture
  of “topics” with different rates or probabilities.</p>
<p>Once the loadings and factors have been generated, the counts are
  simulated from either the Poisson NMF or multinomial topic model:
  for the former, <code>X[i,j]</code> is Poisson with rate <code>Y[i,j]</code>,
  where <code>Y = tcrossprod(L,F)</code>; for the latter, <code>X[i,]</code> is
  multinomial with size <code>s[i]</code> and with class probabilities
  <code>P[i,]</code>, where <code>P = tcrossprod(L,F)</code>. For the multinomial
  model only, the sizes <code>s</code> are randomly generated as <code>s =
  10^rnorm(n,3,0.2)</code>.</p>
<p>Note that only minimal argument checking is performed;
  the function is mainly used to test implementation of the
  topic-model-based differential count analysis.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Peter Carbonetto, Kevin Luo, Kushal Dey, Matthew
Stephens.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a>
2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

