<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Structure Plot — structure_plot • fastTopics</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Structure Plot — structure_plot" />
<meta property="og:description" content="Create a &amp;#8220;Structure plot&amp;#8221; from a multinomial topic
  model fit. The Structure plot represents the estimated topic
  proportions of each sample in a stacked bar chart, with bars of
  different colors representing different topics. Consequently,
  samples that have similar topic proportions have similar amounts
  of each color." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fastTopics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5-52</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/index.html">Vignettes</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/fastTopics">Source</a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Structure Plot</h1>
    <small class="dont-index">Source: <a href='https://github.com/stephenslab/fastTopics/blob/master/R/structure_plot.R'><code>R/structure_plot.R</code></a></small>
    <div class="hidden name"><code>structure_plot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Create a &#8220;Structure plot&#8221; from a multinomial topic
  model fit. The Structure plot represents the estimated topic
  proportions of each sample in a stacked bar chart, with bars of
  different colors representing different topics. Consequently,
  samples that have similar topic proportions have similar amounts
  of each color.</p>
    </div>

    <pre class="usage"><span class='fu'>structure_plot</span>(
  <span class='no'>fit</span>,
  <span class='no'>grouping</span>,
  <span class='no'>topics</span>,
  <span class='no'>rows</span>,
  <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>2000</span>,
  <span class='kw'>colors</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"#e41a1c"</span>, <span class='st'>"#377eb8"</span>, <span class='st'>"#4daf4a"</span>, <span class='st'>"#984ea3"</span>, <span class='st'>"#ff7f00"</span>, <span class='st'>"#ffff33"</span>,
    <span class='st'>"#a65628"</span>, <span class='st'>"#f781bf"</span>, <span class='st'>"#999999"</span>),
  <span class='kw'>gap</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>embed_method</span> <span class='kw'>=</span> <span class='no'>structure_plot_default_embed_method</span>,
  <span class='kw'>ggplot_call</span> <span class='kw'>=</span> <span class='no'>structure_plot_ggplot_call</span>,
  <span class='no'>...</span>
)

<span class='fu'>structure_plot_default_embed_method</span>(<span class='no'>fit</span>, <span class='no'>...</span>)

<span class='co'># S3 method for poisson_nmf_fit</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.html'>plot</a></span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='co'># S3 method for multinom_topic_model_fit</span>
<span class='fu'><a href='https://rdrr.io/r/graphics/plot.html'>plot</a></span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='fu'>structure_plot_ggplot_call</span>(<span class='no'>dat</span>, <span class='no'>colors</span>, <span class='kw'>ticks</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>font.size</span> <span class='kw'>=</span> <span class='fl'>9</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>fit</th>
      <td><p>An object of class &#8220;poisson_nmf_fit&#8221; or
&#8220;multinom_topic_model_fit&#8221;. If a Poisson NMF fit is provided
as input, the corresponding multinomial topic model fit is
automatically recovered using <code><a href='poisson2multinom.html'>poisson2multinom</a></code>.</p></td>
    </tr>
    <tr>
      <th>grouping</th>
      <td><p>Optional categorical variable (a factor) with one
entry for each row of the loadings matrix <code>fit$L</code> defining a
grouping of the samples (rows). The samples (rows) are arranged
along the horizontal axis according to this grouping, then within
each group according to <code>rows</code> or, if not provided, according
to the 1-d embedding computed by <code>embed_method</code>. If
<code>grouping</code> is not a factor, an attempt is made to convert it
to a factor using <code>as.factor</code>.</p></td>
    </tr>
    <tr>
      <th>topics</th>
      <td><p>Top-to-bottom ordering of the topics in the Structure
plot; <code>topics[1]</code> is shown on the top, <code>topics[2]</code> is
shown next, and so on. If the ordering of the topics is not
specified, the topics are automatically ordered so that the topics
with the greatest total &#8220;mass&#8221; are at shown at the bottom of
the plot. The topics may be specified by number or by name.</p></td>
    </tr>
    <tr>
      <th>rows</th>
      <td><p>Ordering of the rows (samples) in the Structure plot
after they have been grouped. This may include all the rows, or a
subset of the rows. If not provided, it is generated automatically
from a 1-d embedding, separately for each group. The rows may be
specified by number or by name.</p></td>
    </tr>
    <tr>
      <th>n</th>
      <td><p>The maximum number of samples (rows of the loadings matrix
<code>fit$L</code>) to include in the plot. Typically there is little to
no benefit in including large number of samples in the Structure
plot due to screen resolution limits. Ignored if <code>rows</code> is
provided.</p></td>
    </tr>
    <tr>
      <th>colors</th>
      <td><p>Colors used to draw topics in Structure plot. The
default colour setting is the from <a href='https://colorbrewer2.org'>https://colorbrewer2.org</a>
(qualitative data, &#8220;9-class Set1&#8221;).</p></td>
    </tr>
    <tr>
      <th>gap</th>
      <td><p>The horizontal spacing between groups. Ignored if
<code>grouping</code> is not provided.</p></td>
    </tr>
    <tr>
      <th>embed_method</th>
      <td><p>The function used to compute an 1-d embedding
from a loadings matrix <code>fit$L</code>; ignored if <code>rows</code> is
provided. The function must accept the multinomial topic model fit
as its first input (&#8220;fit&#8221;) and additional arguments may be
passed (...). The output should be a named numeric vector with
one entry per row of <code>fit$L</code>, and the names of the entries
should be the same as the row names of <code>fit$L</code>.</p></td>
    </tr>
    <tr>
      <th>ggplot_call</th>
      <td><p>The function used to create the plot. Replace
<code>structure_plot_ggplot_call</code> with your own function to
customize the appearance of the plot.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments passed to <code>structure_plot</code>
(for the <code>plot</code> method) or <code>embed_method</code> (for
function <code>structure_plot</code>).</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>An object of class &#8220;poisson_nmf_fit&#8221; or
&#8220;multinom_topic_model_fit&#8221;. If a Poisson NMF fit is provided
as input, the corresponding multinomial topic model fit is
automatically recovered using <code><a href='poisson2multinom.html'>poisson2multinom</a></code>.</p></td>
    </tr>
    <tr>
      <th>dat</th>
      <td><p>A data frame passed as input to
<code><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></code>, containing, at a minimum, columns
&#8220;sample&#8221;, &#8220;topic&#8221; and &#8220;prop&#8221;: the
&#8220;sample&#8221; column contains the positions of the samples (rows
of the L matrix) along the horizontal axis; the &#8220;topic&#8221;
column is a topic (a column of L); and the &#8220;prop&#8221; column is
the topic proportion for the respective sample.</p></td>
    </tr>
    <tr>
      <th>ticks</th>
      <td><p>The placement of the group labels along the horizontal
axis, and their names. For data that are not grouped, use
<code>ticks = NULL</code>.</p></td>
    </tr>
    <tr>
      <th>font.size</th>
      <td><p>Font size used in plot.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>ggplot</code> object.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The name &#8220;Structure plot&#8221; comes from its widespread
use in population genetics to visualize the results of the
Structure software (Rosenberg <em>et al</em>, 2002).</p>
<p>For most uses of the Structure plot in population genetics, there
is usually some grouping of the samples (e.g., assignment to
pre-defined populations) that guides arrangement of the samples
along the horizontal axis in the bar chart. In other applications,
such as analysis of gene expression data, a pre-defined grouping
may not always be available. Therefore, a &#8220;smart&#8221;
arrangement of the samples is, by default, generated automatically
by performing a 1-d embedding of the samples.</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Dey, K. K., Hsiao, C. J. and Stephens, M. (2017). Visualizing the
structure of RNA-seq expression data using grade of membership
models. <em>PLoS Genetics</em> <b>13</b>, e1006599.</p>
<p>Rosenberg, N. A., Pritchard, J. K., Weber, J. L., Cann, H. M.,
Kidd, K. K., Zhivotovsky, L. A. and Feldman, M. W. (2002). Genetic
structure of human populations. <em>Science</em> <b>298</b>,
2381–2385.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span>(<span class='fl'>1</span>)
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>pbmc_facs</span>)

<span class='co'># Get the multinomial topic model fitted to the</span>
<span class='co'># PBMC data.</span>
<span class='no'>fit</span> <span class='kw'>&lt;-</span> <span class='no'>pbmc_facs</span>$<span class='no'>fit</span>

<span class='co'># Create a Structure plot without labels. The samples (rows) are</span>
<span class='co'># automatically arranged along the x-axis using t-SNE to highlight</span>
<span class='co'># the structure in the data.</span>
<span class='no'>p1</span> <span class='kw'>&lt;-</span> <span class='fu'>structure_plot</span>(<span class='no'>fit</span>)</div><div class='output co'>#&gt; <span class='message'>Running tsne on 2000 x 6 matrix.</span></div><div class='input'>
<span class='co'># Create a Structure plot with the FACS cell-type labels. Within each</span>
<span class='co'># group (cell-type), the cells (rows) are automatically arranged using</span>
<span class='co'># t-SNE.</span>
<span class='no'>subpop</span> <span class='kw'>&lt;-</span> <span class='no'>pbmc_facs</span>$<span class='no'>samples</span>$<span class='no'>subpop</span>
<span class='no'>p2</span> <span class='kw'>&lt;-</span> <span class='fu'>structure_plot</span>(<span class='no'>fit</span>,<span class='kw'>grouping</span> <span class='kw'>=</span> <span class='no'>subpop</span>)</div><div class='output co'>#&gt; <span class='message'>Running tsne on 428 x 6 matrix.</span></div><div class='output co'>#&gt; <span class='message'>Running tsne on 89 x 6 matrix.</span></div><div class='output co'>#&gt; <span class='message'>Running tsne on 337 x 6 matrix.</span></div><div class='output co'>#&gt; <span class='message'>Running tsne on 356 x 6 matrix.</span></div><div class='output co'>#&gt; <span class='message'>Running tsne on 790 x 6 matrix.</span></div><div class='input'>
<span class='co'># Next, we apply some customizations to improve the plot: (1) use the</span>
<span class='co'># "topics" argument to specify the order in which the topic</span>
<span class='co'># proportions are stacked on top of each other; (2) use the "gap"</span>
<span class='co'># argrument to increase the whitespace between the groups; (3) use "n"</span>
<span class='co'># to decrease the number of rows included in the Structure plot; and</span>
<span class='co'># (4) use "colors" to change the colors used to draw the topic</span>
<span class='co'># proportions.</span>
<span class='no'>topic_colors</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"skyblue"</span>,<span class='st'>"forestgreen"</span>,<span class='st'>"darkmagenta"</span>,
                  <span class='st'>"dodgerblue"</span>,<span class='st'>"gold"</span>,<span class='st'>"darkorange"</span>)
<span class='no'>p3</span> <span class='kw'>&lt;-</span> <span class='fu'>structure_plot</span>(<span class='no'>fit</span>,<span class='kw'>grouping</span> <span class='kw'>=</span> <span class='no'>pbmc_facs</span>$<span class='no'>samples</span>$<span class='no'>subpop</span>,<span class='kw'>gap</span> <span class='kw'>=</span> <span class='fl'>20</span>,
                     <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>1500</span>,<span class='kw'>topics</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>5</span>,<span class='fl'>6</span>,<span class='fl'>1</span>,<span class='fl'>4</span>,<span class='fl'>2</span>,<span class='fl'>3</span>),<span class='kw'>colors</span> <span class='kw'>=</span> <span class='no'>topic_colors</span>)</div><div class='output co'>#&gt; <span class='message'>Running tsne on 295 x 6 matrix.</span></div><div class='output co'>#&gt; <span class='message'>Running tsne on 72 x 6 matrix.</span></div><div class='output co'>#&gt; <span class='message'>Running tsne on 287 x 6 matrix.</span></div><div class='output co'>#&gt; <span class='message'>Running tsne on 264 x 6 matrix.</span></div><div class='output co'>#&gt; <span class='message'>Running tsne on 582 x 6 matrix.</span></div><div class='input'>
<span class='co'># In this final example, we use UMAP instead of t-SNE to arrange all</span>
<span class='co'># 3,744 cells in the Structure plot. Note that this can be</span>
<span class='co'># accomplished in a different way by overriding the default setting of</span>
<span class='co'># "embed_method".</span>
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/drop.html'>drop</a></span>(<span class='fu'><a href='embeddings_from_topics.html'>umap_from_topics</a></span>(<span class='no'>fit</span>,<span class='kw'>dims</span> <span class='kw'>=</span> <span class='fl'>1</span>))</div><div class='output co'>#&gt; <span class='message'>10:48:18 UMAP embedding parameters a = 1.896 b = 0.8006</span></div><div class='output co'>#&gt; <span class='message'>10:48:18 Read 3774 rows and found 6 numeric columns</span></div><div class='output co'>#&gt; <span class='message'>10:48:18 Using FNN for neighbor search, n_neighbors = 30</span></div><div class='output co'>#&gt; <span class='message'>10:48:18 Commencing smooth kNN distance calibration using 2 threads</span></div><div class='output co'>#&gt; <span class='message'>10:48:19 111 smooth knn distance failures</span></div><div class='output co'>#&gt; <span class='message'>10:48:19 Initializing from normalized Laplacian + noise</span></div><div class='output co'>#&gt; <span class='message'>10:48:20 Commencing optimization for 500 epochs, with 138012 positive edges</span></div><div class='output co'>#&gt; <span class='message'>10:48:25 Optimization finished</span></div><div class='input'><span class='no'>p4</span> <span class='kw'>&lt;-</span> <span class='fu'>structure_plot</span>(<span class='no'>fit</span>,<span class='kw'>rows</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/order.html'>order</a></span>(<span class='no'>y</span>),<span class='kw'>grouping</span> <span class='kw'>=</span> <span class='no'>subpop</span>,<span class='kw'>gap</span> <span class='kw'>=</span> <span class='fl'>40</span>,
                     <span class='kw'>colors</span> <span class='kw'>=</span> <span class='no'>topic_colors</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Peter Carbonetto, Kevin Luo, Kushal Dey, Matthew Stephens.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


